Build both Android and iOS bundles in Buildkite CI

mokagio opened the pull request:
Building the iOS JS bundle achieves two goals:

- Allows the step that builds the XCFramework to add the bundle as a resource without spending time setting up all the node stack. This is efficient because we make use of the work that has already run for the Android bundle
- Brings us one step closer to being able to run all our CI tasks in Buildkite over Circle CI

_These were originally part of #5769 but I keep running into last mile issues to the get generated and downloaded XCFramework to work, so I thought I'd track these orthogonal changes in a dedicated PR._

To test: Verify build was successful in Buildkite

<img width="1163" alt="image" src="https://github.com/wordpress-mobile/gutenberg-mobile/assets/1218433/27556747-fee5-4e7c-8e5b-427e9724579d">

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered if this change warrants user-facing release notes [more info](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/Release-notes.md) and have added them to [RELEASE-NOTES.txt](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/RELEASE-NOTES.txt) if necessary.

mokagio committed:
Update "Bundle Android" CI step to also build iOS bundle

And accordingly rename it to "Build JS bundles"

mokagio committed:
Organize the "Build JS bundles" CI step logs with Buildkite groups

mokagio committed:
Remove unnecessary Homebrew environment definitions for iOS step

The flags we need are already part of the VM image 🎉

mokagio committed:
Use the Xcode 14.3 image to build the XCFramework in CI

Just to stay up to date with the latest tooling

mokagio commented on .buildkite/pipeline.yml:
```diff
@@ -23,34 +23,54 @@ steps:
+        echo "--- :npm: Install Node dependencies"
         npm ci --unsafe-perm --prefer-offline --no-audit --no-progress
 
+        echo "--- :package: Run bundle prep work"
         npm run prebundle:js
```
For reference and future optimization avenues, this is the command that takes up the bulk of the execution time.

mokagio commented on .buildkite/publish-react-native-ios-artifacts.sh:
```diff
@@ -1,16 +1,22 @@
 #!/bin/bash -eu
 
-cd ./ios-xcframework
+# Make Homebrew run a bit faster
+export HOMEBREW_NO_AUTO_UPDATE=1
+export HOMEBREW_NO_INSTALL_CLEANUP=1
```
Having these exports here rather than inline is mostly cosmetic. I did it because I was `brew install`ing another tool at some point in development, but that's no longer the case.

I considered using the `env` node in the step definition in `pipeline.yml` but I wasn't sure whether we propagate those values to the scripts we run. Besides, given the code that uses these var is in this file, it make sense to me to leave them here.

mokagio replied:
Also notice it's been discussed (internal ref p1682611494399739-slack-C020Q0Z579V ) adding these to our VM setup upstream. When that will happen, we'll be able to remove these from here.

crazytonyli replied:
`HOMEBREW_NO_AUTO_UPDATE` is already included. `HOMEBREW_NO_INSTALL_CLEANUP` is not. But I don't think we need the second one? I don't imagine "cleaning up old formula" would have an big impact on the performance.

> HOMEBREW_NO_INSTALL_CLEANUP
>
> If set, brew install, brew upgrade and brew reinstall will never automatically cleanup installed/upgraded/reinstalled formulae or all formulae every HOMEBREW_CLEANUP_PERIODIC_FULL_DAYS days. Alternatively, HOMEBREW_NO_CLEANUP_FORMULAE allows specifying specific formulae to not clean up.

mokagio commented on .buildkite/pipeline.yml:
```diff
@@ -23,34 +23,54 @@ steps:
             # Allow WP-CLI to be run as root, otherwise it throws an exception.
             # Reference: https://git.io/J9q2S
             - "WP_CLI_ALLOW_ROOT=true"
+      - peakon/git-shallow-clone#v0.0.1:
+          depth: 1
```
This plugin makes checking out the repo ~ 2 minutes faster

**Before ([build](https://buildkite.com/automattic/gutenberg-mobile/builds/5869#018855c3-bc85-45cb-93be-41386965a05d))**

![image](https://github.com/wordpress-mobile/gutenberg-mobile/assets/1218433/9ba460f9-ab93-41c1-afb4-eeb6f4e39212)


**After ([build](https://buildkite.com/automattic/gutenberg-mobile/builds/5870#0188567c-9da5-4726-85bd-a2b5d132a314))**

<img width="1148" alt="image" src="https://github.com/wordpress-mobile/gutenberg-mobile/assets/1218433/bd751d00-a3e4-44d4-9d03-c4ce9d5909f8">

mokagio replied:
🤔 I guess I could DRY this by defining it in `x-common-params`.

I was hoping there was a way to add a plugin to every step, but [the docs don't mention it](https://buildkite.com/docs/plugins/using#using-yaml-anchors-with-plugins).

crazytonyli commented on .buildkite/pipeline.yml:
```diff
@@ -23,42 +26,58 @@ steps:
       queue: mac
     env:
-      IMAGE_ID: xcode-14.2
+      IMAGE_ID: xcode-14.3
```
Considering we are updating Xcode, may as well update it to the latest one `14.3.1-rc.1` which is available on CI. 😸

mokagio replied:
During development, I experienced an issue which was resolved, I think, by putting the builder and consumer projects on the same Xcode version 😳 

For the moment, I'd like to keep the two in sync.

mokagio commented:
@crazytonyli thank you so much for the review on this. I should have addressed all your comments. I did that via a rebase to keep the history neat but it came at the downside of having to re-review the entire diff.

oguzkocer reviewed (approved):
Looks good to me :shipit:
